presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/trait/CardinalityTraitCalculationRuleSet.java

```diff
@@ -0,0 +1,132 @@
+import static com.facebook.presto.sql.planner.plan.Patterns.limit;
+import static com.facebook.presto.sql.planner.plan.Patterns.markDistinct;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.values;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.Math.min;
+
+public final class CardinalityTraitCalculationRuleSet
+        implements RuleSet
+{
+    @Override
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(
+                simpleCardinalityRule(enforceSingleRow(), (node, context) -> scalar()),
+                simpleCardinalityRule(values(), (node, context) -> exactly(node.getRows().size())),
+                simpleCardinalityRule(
+                        aggregation(),
+                        (node, context) -> node.getStep() == SINGLE && node.getGroupingSets().size() == 1 && node.hasEmptyGroupingSet(),
+                        (node, context) -> scalar()),
+                propagateTrait(project()),
+                propagateTrait(markDistinct()),
+                propagateTrait(exchange(), (node, context) -> node.getSources().size() == 1),
+                simpleCardinalityRule(filter(), CardinalityTraitCalculationRuleSet::sourceHasKnownCardinality, (node, context) -> {
+                    CardinalityTrait sourceCardinality = context.getLookup().resolveTrait(node.getSource(), CARDINALITY).get();
+                    return atMost(sourceCardinality.getCardinalityRange().upperEndpoint());
+                }),
+                simpleCardinalityRule(
+                        limit(),
+                        (node, context) -> {
+                            Optional<CardinalityTrait> sourceCardinality = context.getLookup().resolveTrait(node.getSource(), CARDINALITY);
+                            long upper = node.getCount();
+                            long lower = 0;
+                            if (sourceCardinality.isPresent()) {
+                                upper = min(sourceCardinality.get().getCardinalityRange().upperEndpoint(), upper);
+                                lower = min(upper, sourceCardinality.get().getCardinalityRange().lowerEndpoint());
+                            }
+                            return new CardinalityTrait(Range.closed(lower, upper));
+                        }));
+    }
+
+    private <T extends PlanNode> Rule<T> propagateTrait(Pattern<T> pattern)
+    {
+        return propagateTrait(pattern, alwaysTrue());
+    }
+
+    private <T extends PlanNode> Rule<T> propagateTrait(Pattern<T> pattern, CardinalityDerivePredicate<T> predicate)
+    {
+        return simpleCardinalityRule(
+                pattern,
+                (node, context) -> predicate.apply(node, context) && sourceHasKnownCardinality(node, context),
+                (node, context) -> context.getLookup().resolveTrait(getOnlyElement(node.getSources()), CARDINALITY).get());
```

Comment:
`propagateTrait` resolves trait twice even though it could do it once if implemented in a more traditional fasion